helix-term/src/commands/engine/scheme.rs

```diff
@@ -0,0 +1,2003 @@
+use helix_core::{
+    extensions::steel_implementations::{rope_module, SteelRopeSlice},
+    graphemes,
+    path::expand_tilde,
+    regex::Regex,
+    shellwords::Shellwords,
+    syntax::Configuration,
+    Range, Selection, Tendril,
+};
+use helix_event::register_hook;
+use helix_loader::merge_toml_values;
+use helix_lsp::lsp::CompletionItem;
+use helix_view::{
+    document::Mode,
+    editor::{Action, ConfigEvent},
+    extension::document_id_to_usize,
+    input::KeyEvent,
+    Document, DocumentId, Editor, Theme,
+};
+use once_cell::sync::Lazy;
+use serde_json::Value;
+use steel::{
+    gc::unsafe_erased_pointers::CustomReference,
+    rerrs::ErrorKind,
+    rvals::{as_underlying_type, AsRefMutSteelValFromRef, FromSteelVal, IntoSteelVal},
+    steel_vm::{engine::Engine, register_fn::RegisterFn},
+    steelerr, SteelErr, SteelVal,
+};
+
+use std::{borrow::Cow, cell::RefCell, collections::HashMap, ops::Deref, path::PathBuf, rc::Rc};
+use std::{
+    collections::HashSet,
+    sync::{Arc, RwLock},
+};
+
+use steel::{rvals::Custom, steel_vm::builtin::BuiltInModule};
+
+use crate::{
+    commands::insert,
+    compositor::{self, Component, Compositor},
+    config::Config,
+    events::{OnModeSwitch, PostCommand, PostInsertChar},
+    job::{self, Callback},
+    keymap::{self, merge_keys, KeyTrie, KeymapResult},
+    ui::{self, Popup, Prompt, PromptEvent},
+};
+
+use components::SteelDynamicComponent;
+
+use super::{components, shell_impl, Context, MappableCommand, TYPABLE_COMMAND_LIST};
+use insert::{insert_char, insert_string};
+
+thread_local! {
+    pub static ENGINE: std::rc::Rc<std::cell::RefCell<steel::steel_vm::engine::Engine>> = configure_engine();
+}
+
+enum CoreModules {
+    Document(DocumentApi),
+    Editor(EditorApi),
+    Component(ComponentApi),
+    TypedCommands(TypedCommandsApi),
+    StaticCommands(StaticCommandsApi),
+}
+
+// APIs / Modules that need to be accepted by the plugin system
+// Without these, the core functionality cannot operate
+pub struct DocumentApi;
+pub struct EditorApi;
+pub struct ComponentApi;
+pub struct TypedCommandsApi;
+pub struct StaticCommandsApi;
+
+pub struct KeyMapApi {
+    get_keymap: fn() -> EmbeddedKeyMap,
+    default_keymap: fn() -> EmbeddedKeyMap,
+    empty_keymap: fn() -> EmbeddedKeyMap,
+    string_to_embedded_keymap: fn(String) -> EmbeddedKeyMap,
+    merge_keybindings: fn(&mut EmbeddedKeyMap, EmbeddedKeyMap),
+    is_keymap: fn(SteelVal) -> bool,
+    deep_copy_keymap: fn(EmbeddedKeyMap) -> EmbeddedKeyMap,
+}
+
+impl KeyMapApi {
+    fn new() -> Self {
+        KeyMapApi {
+            get_keymap,
+            default_keymap,
+            empty_keymap,
+            string_to_embedded_keymap,
+            merge_keybindings,
+            is_keymap,
+            deep_copy_keymap,
+        }
+    }
+}
+
+thread_local! {
```

Comment:
I am not sure how suitable threadlocals really are. 

I think tokio can and does move our main loop to different threads (and I also don't want to constrain interaction with plugins to a single thread).

In the worst case these could be globals but for almost all of these it feels like its really adding more extensability to the editor here instead of in the editor itself.


I think we should instead take the approach of making the core editor more extensible and then exposing those APIs to plugins.

These APIs would ofcourse require access to somekind of context. This would be similar to the context problem you mentinoed in your description. I think the best way to handle that one would be if we stored the context as a global inside the engine (using run_with_reference).

In any callbacks that want access to the context we check for that global (even user code could do it themselves right?) and throw an error if a function that needs a context is called without context access.

Reply:
I'm currently refactoring the implementation to instead just from the steel side of things, have one variable that refers to the context. Whenever we call steel code, we'll do like you said - just refresh the context variable with the current context for the duration of the steel code scope, otherwise it'll be some poisoned value that will error.

With respect to thread locals - I'm really not sure how else to implement the interaction with the steel engine if not through a thread local, and any callbacks that I set up are done on a thread local queue that tokio is processing.

Can you elaborate on:
> I also don't want to constrain interaction with plugins to a single thread

Does this mean you want the engine to be utilizing multiple threads? Or you want the main loop to be able to be moved to different threads (or both)?

As it stands, the engine implementation is not thread safe, which would make this rather tricky to implement. I don't know if tokio is actively moving the main loop to another thread or not, as I haven't run into any strange behavior with the interaction between the editor and the engine up to this point.

Reply:
To put it another way - if we didn't want to have the main thread interact with the scripting engine, then it would have to take the context by value for this to work (at least, that is what I'm currently thinking)

Reply:
If tokio lets you hold the engine struct across await points, I think the main loop only ever runs on a single thread. I would like to be able to eventually make `Engine` `Send` so we can send it across thread. I could definitely imagine that we would like to run some plugin code in the background (or in response to some async event). 

Requiring engine to be `Sync` won't work since you would need to pub mutexes everywhere which would be excessively slow. But Send could be reasonable so we can have `Mutex<Engine>` and share that across threads. That is usually a reasonably easy requirement. The main thing you can't do is expose something like `Rc` to the user. I know you use refcounting internally (so `Send` is currently not implemented) but it could be safe to `unsafe impl Send for Engine` assuming that you never hand any refcounted types out to rust (or if you do make sure to use `Arc` or a non-clonable reference).

I think thread_locals are quite alright for the `Context` since the context would be explicitly provided everytime we call into the `Engine`.

Reply:
So that the current moment, I've added some code to have the whole editor panic if the engine gets used on another thread to see if it is happening. I have been using this as a daily driver for a few weeks like this and haven't yet run into it. So while I agree thread local isn't the best, it also seems to be acting as intended, which I think is just by coincidence